File: src/main/tasks.js
'use strict'
// @ts-check
/**
 * 后台任务系统：
 * - 在主进程维护任务注册表（跨页面持久，不随渲染进程状态丢失）
 * - 任务有生命周期：running → done | error | cancelled
 * - 通过 IPC 向渲染进程推送 'task:update' 事件（增量）
 * - 渲染进程任何时候都能 list / subscribe 拿到全部任务
 *
 * 任务类型：
 *  - generate   AI 报告生成
 *  - memory     AI 记忆生成/重建
 *  - model_dl   模型下载（embedding）
 *  - webdav     WebDAV 同步
 *  - custom     其它
 */

let _tasks = new Map() // id → task object
let _sender = null      // 推送函数：({type, task}) => void
let _seq = 0

function newId() {
  _seq++
  return `task_${Date.now().toString(36)}_${_seq}`
}

/** 注入推送函数（由 ipc.js 注册时注入，转发到渲染进程） */
function setSender(fn) {
  _sender = fn
}

function emit(task) {
  if (_sender) {
    try { _sender({ type: 'update', task: snapshot(task) }) } catch {}
  }
}

/** 生成对外快照（去掉内部字段） */
function snapshot(task) {
  return {
    id: task.id,
    kind: task.kind,
    title: task.title,
    status: task.status, // running | done | error | cancelled
    progress: task.progress, // { done, total, label }
    detail: task.detail,
    error: task.error,
    result: task.result,
    createdAt: task.createdAt,
    updatedAt: task.updatedAt,
  }
}

/**
 * 创建一个任务，返回 { id, update, done, error }
 * 调用方在异步流程中用 update/done/error 推进状态。
 */
function create(kind, title, opts = {}) {
  const id = newId()
  const task = {
    id,
    kind, // generate | memory | model_dl | webdav | custom
    title,
    status: 'running',
    progress: opts.progress || null, // { done, total, label }
    detail: opts.detail || '',
    error: null,
    result: null,
    createdAt: Date.now(),
    updatedAt: Date.now(),
  }
  _tasks.set(id, task)
  // 限制最近 50 条，避免无限增长
  if (_tasks.size > 50) {
    const oldest = _tasks.keys().next().value
    _tasks.delete(oldest)
  }
  emit(task)
  return id
}

/** 更新任务进度 */
function update(id, patch) {
  const task = _tasks.get(id)
  if (!task) return
  if (patch.progress !== undefined) task.progress = patch.progress
  if (patch.detail !== undefined) task.detail = patch.detail
  if (patch.title !== undefined) task.title = patch.title
  task.updatedAt = Date.now()
  emit(task)
}

/** 标记任务完成 */
function done(id, result) {
  const task = _tasks.get(id)
  if (!task) return
  task.status = 'done'
  task.result = result || null
  task.updatedAt = Date.now()
  emit(task)
}

/** 标记任务失败 */
function error(id, errorMsg) {
  const task = _tasks.get(id)
  if (!task) return
  task.status = 'error'
  task.error = errorMsg || '未知错误'
  task.updatedAt = Date.now()
  emit(task)
}

/** 取消（标记为 cancelled，不强制中断） */
function cancel(id) {
  const task = _tasks.get(id)
  if (!task) return
  task.status = 'cancelled'
  task.updatedAt = Date.now()
  emit(task)
}

/** 删除一条任务记录 */
function remove(id) {
  _tasks.delete(id)
  if (_sender) { try { _sender({ type: 'remove', id }) } catch {} }
}

/** 清除已完成/失败的任务 */
function clearFinished() {
  const toRemove = []
  for (const [id, task] of _tasks) {
    if (task.status !== 'running') toRemove.push(id)
  }
  toRemove.forEach((id) => _tasks.delete(id))
  if (_sender) { try { _sender({ type: 'clear' }) } catch {} }
}

/** 获取所有任务快照（按创建时间倒序） */
function list() {
  return [..._tasks.values()]
    .sort((a, b) => b.createdAt - a.createdAt)
    .map(snapshot)
}

/** 当前是否有运行中的任务 */
function hasRunning() {
  for (const task of _tasks.values()) {
    if (task.status === 'running') return true
  }
  return false
}

module.exports = {
  setSender,
  create,
  update,
  done,
  error,
  cancel,
  remove,
  clearFinished,
  list,
  hasRunning,
}

File: src-tauri/src/tasks.rs
//! 后台任务系统：对齐 src/main/tasks.js。
//! 在主进程维护任务注册表（跨页面持久），生命周期 running → done|error|cancelled，
//! 经 `task:update` 事件增量推送到渲染层。用 Tauri State 持有，AppHandle 负责 emit。

use std::sync::Mutex;

use serde::Serialize;
use serde_json::Value;
use tauri::{AppHandle, Emitter};

use crate::utils;

const MAX_TASKS: usize = 50;

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Progress {
    pub done: f64,
    pub total: f64,
    pub label: String,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Task {
    pub id: String,
    pub kind: String,
    pub title: String,
    pub status: String, // running | done | error | cancelled
    pub progress: Option<Progress>,
    pub detail: String,
    pub error: Option<String>,
    pub result: Value,
    pub created_at: u128,
    pub updated_at: u128,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct UpdatePayload<'a> {
    #[serde(rename = "type")]
    kind: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    task: Option<&'a Task>,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<&'a str>,
}

#[derive(Default)]
struct Inner {
    tasks: Vec<Task>,
    seq: u64,
    app: Option<AppHandle>,
}

/// 任务注册表（作为 Tauri managed state）。
#[derive(Default)]
pub struct Tasks {
    inner: Mutex<Inner>,
}

fn emit_update(app: &AppHandle, task: &Task) {
    let _ = app.emit(
        "task:update",
        UpdatePayload {
            kind: "update",
            task: Some(task),
            id: None,
        },
    );
}

impl Tasks {
    pub fn new() -> Self {
        Tasks::default()
    }

    /// 注入 AppHandle（在 setup 中调用，使后续任务变更可推送到前端）。
    pub fn set_app(&self, app: AppHandle) {
        self.inner.lock().unwrap().app = Some(app);
    }

    /// 创建任务，返回 id（对齐 tasks.js create）。
    pub fn create(&self, kind: &str, title: &str, detail: &str, progress: Option<Progress>) -> String {
        let (snapshot, app) = {
            let mut g = self.inner.lock().unwrap();
            g.seq += 1;
            let id = format!("task_{}_{}", utils::base36(utils::now_ms()), g.seq);
            let now = utils::now_ms();
            let task = Task {
                id: id.clone(),
                kind: kind.to_string(),
                title: title.to_string(),
                status: "running".to_string(),
                progress,
                detail: detail.to_string(),
                error: None,
                result: Value::Null,
                created_at: now,
                updated_at: now,
            };
            g.tasks.push(task.clone());
            // 限制最近 MAX_TASKS 条（删最老）
            if g.tasks.len() > MAX_TASKS {
                g.tasks.remove(0);
            }
            (task, g.app.clone())
        };
        if let Some(app) = app {
            emit_update(&app, &snapshot);
        }
        snapshot.id
    }

    fn mutate<F: FnOnce(&mut Task)>(&self, id: &str, f: F) {
        let (snapshot, app) = {
            let mut g = self.inner.lock().unwrap();
            let Some(task) = g.tasks.iter_mut().find(|t| t.id == id) else {
                return;
            };
            f(task);
            task.updated_at = utils::now_ms();
            (task.clone(), g.app.clone())
        };
        if let Some(app) = app {
            emit_update(&app, &snapshot);
        }
    }

    /// 更新进度/详情/标题（对齐 tasks.js update）。
    pub fn update(
        &self,
        id: &str,
        detail: Option<String>,
        progress: Option<Option<Progress>>,
        title: Option<String>,
    ) {
        self.mutate(id, |task| {
            if let Some(p) = progress {
                task.progress = p;
            }
            if let Some(d) = detail {
                task.detail = d;
            }
            if let Some(t) = title {
                task.title = t;
            }
        });
    }

    /// 标记完成（对齐 tasks.js done）。
    pub fn done(&self, id: &str, result: Value) {
        self.mutate(id, |task| {
            task.status = "done".to_string();
            task.result = result;
        });
    }

    /// 标记失败（对齐 tasks.js error）。
    pub fn error(&self, id: &str, error_msg: &str) {
        self.mutate(id, |task| {
            task.status = "error".to_string();
            task.error = Some(if error_msg.is_empty() {
                "未知错误".to_string()
            } else {
                error_msg.to_string()
            });
        });
    }

    /// 删除一条任务（对齐 tasks.js remove）。
    pub fn remove(&self, id: &str) {
        let app = {
            let mut g = self.inner.lock().unwrap();
            g.tasks.retain(|t| t.id != id);
            g.app.clone()
        };
        if let Some(app) = app {
            let _ = app.emit(
                "task:update",
                UpdatePayload {
                    kind: "remove",
                    task: None,
                    id: Some(id),
                },
            );
        }
    }

    /// 清除已完成/失败的任务（对齐 tasks.js clearFinished）。
    pub fn clear_finished(&self) {
        let app = {
            let mut g = self.inner.lock().unwrap();
            g.tasks.retain(|t| t.status == "running");
            g.app.clone()
        };
        if let Some(app) = app {
            let _ = app.emit(
                "task:update",
                UpdatePayload {
                    kind: "clear",
                    task: None,
                    id: None,
                },
            );
        }
    }

    /// 全部任务快照（按创建时间倒序，对齐 tasks.js list）。
    pub fn list(&self) -> Vec<Task> {
        let g = self.inner.lock().unwrap();
        let mut out = g.tasks.clone();
        out.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        out
    }

    /// 是否有运行中的任务（对齐 tasks.js hasRunning）。
    pub fn has_running(&self) -> bool {
        self.inner
            .lock()
            .unwrap()
            .tasks
            .iter()
            .any(|t| t.status == "running")
    }
}
